pdl_syntax.md

```diff
@@ -0,0 +1,828 @@
+```
+
+The dot seperate keys format is equivalent to JSON value format.
+
+The above dot seperate keys example is equivalent to the following JSON value example :
+```
+@java = {
+  "class": "org.joda.DateTime",
+  "coercerClass": "org.example.DateTimeCoercer"
+}
+```
+
+### Escaping
+There are some keywords which are reserved in Pegasus. If you have to use them to define any names, you need to put them in backticks : ` `.
+
+#### Keyword Escaping
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.escaping
+
+record PdlKeywordEscaping {
+  `namespace`: string
+  `record`: string
+  `null`: string
+  `enum`: string
+  recordName: record `record` { }
+}
+```
+
+#### Namespace/Package escaping
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.escaping.`record`
+package com.linkedin.pegasus.generator.test.idl.escaping.override.`typeref`
+
+/**
+ * Ensures that the namespace and package are properly escaped at the root as well as in scoped named-type declarations.
+ */
+record NamespacePackageEscaping {
+  x: {
+    namespace com.x.y.z.`enum`
+    package com.a.b.c.`fixed`
+
+    record Foo {}
+  }
+}
+```
+
+#### Property key escaping 
+
+If you want Pegasus to treat property key name with dots as one string key, please use backticks to escape such string.
+For example : 
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.escaping
+
+record PropertyKeyEscaping {
+  @`namespace` = "foo.bar"
+  @`test.path` = 1
+  @validate.`com.linkedin.CustomValidator` = "foo"
+  aField: string
+}
+```
+
+
+## Primitive Types
+
+The Pegasus primitive types are : int, long, float, double, boolean, string and bytes.
+
+For Example :
+```
+namespace com.linkedin.pegasus.generator.test.idl.records
+
+record WithPrimitives {
+  intField: int
+  longField: long
+  floatField: float
+  doubleField: double
+  booleanField: boolean
+  stringField: string
+  bytesField: bytes
+}
+```
+
+## Array Type
+
+Pegesus Arrays are defined within a items type.
+
+For Example : 
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+record WithPrimitivesArray {
+  ints: array[int]
+  longs: array[long]
+  floats: array[float]
+  doubles: array[double]
+  booleans: array[boolean]
+  strings: array[string]
+  bytes: array[bytes]
+}
+```
+
+Record Arrays :
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+
+record WithRecordArray {
+  empties: array[Empty]
+  fruits: array[Fruits]
+}
+```
+
+
+## Map Type
+
+Pegasus maps are defined with a values type and an optional key type.
+
+For example : 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMap {
+  ints: map[string, int]
+  longs: map[string, long]
+  floats: map[string, float]
+  doubles: map[string, double]
+  booleans: map[string, boolean]
+  strings: map[string, string]
+  bytes: map[string, bytes]
+}
+```
+Custom Types Map :
```

Comment:
This can be removed, not really adding value

Reply:
removed